lib/evse_security/evse_security.cpp

```diff
@@ -1166,122 +1239,154 @@ GetCertificateInfoResult EvseSecurity::get_leaf_certificate_info_internal(LeafCe
+                            << conversions::leaf_certificate_type_to_string(certificate_type);
+                // Move onto next valid leaf, and attempt a search there
+                continue;
             }
 
-            // Found both, break
-            if (leaf_fullchain != nullptr && leaf_single != nullptr)
-                return false;
-
-            return true;
-        });
-
-        std::vector<CertificateOCSP> certificate_ocsp{};
-
-        // None were found
-        if (leaf_single == nullptr && leaf_fullchain == nullptr) {
-            EVLOG_error << "Could not find any leaf certificate for:"
-                        << conversions::leaf_certificate_type_to_string(certificate_type);
+            if (leaf_fullchain != nullptr) {
+                chain_file = leaf_fullchain->at(0).get_file();
+                EVLOG_debug << "Leaf fullchain: [" << chain_file.value_or("INVALID") << "]";
+            } else {
+                EVLOG_debug << conversions::leaf_certificate_type_to_string(certificate_type)
+                            << " leaf requires full bundle, but full bundle not found at path: " << cert_dir;
+            }
 
-            result.status = GetCertificateInfoStatus::NotFound;
-            return result;
-        }
+            if (leaf_single != nullptr) {
+                certificate_file = leaf_single->at(0).get_file();
+                EVLOG_debug << "Leaf single: [" << certificate_file.value_or("INVALID") << "]";
+            } else {
+                EVLOG_debug << conversions::leaf_certificate_type_to_string(certificate_type)
+                            << " single leaf not found at path: " << cert_dir;
+            }
 
-        if (leaf_fullchain != nullptr) {
-            chain_file = leaf_fullchain->at(0).get_file();
-            EVLOG_debug << "Leaf fullchain: [" << chain_file.value_or("INVALID") << "]";
-        } else {
-            EVLOG_warning << conversions::leaf_certificate_type_to_string(certificate_type)
-                          << " leaf requires full bundle, but full bundle not found at path: " << cert_dir;
-        }
+            // Both require the hierarchy build
+            if (include_ocsp || include_root) {
+                X509CertificateBundle root_bundle(root_dir, EncodingFormat::PEM); // Required for hierarchy
 
-        if (leaf_single != nullptr) {
-            certificate_file = leaf_single->at(0).get_file();
-            EVLOG_debug << "Leaf single: [" << certificate_file.value_or("INVALID") << "]";
-        } else {
-            EVLOG_warning << conversions::leaf_certificate_type_to_string(certificate_type)
-                          << " single leaf not found at path: " << cert_dir;
-        }
+                // The hierarchy is required for both roots and the OCSP cache
+                auto hierarchy =
+                    std::move(X509CertificateHierarchy::build_hierarchy(root_bundle.split(), leaf_directory.split()));
```

Comment:
clang logs the following warning:
```
../everest/libevse-security/lib/evse_security/evse_security.cpp:1325:21: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
 1325 |                     std::move(X509CertificateHierarchy::build_hierarchy(root_bundle.split(), leaf_directory.split()));
```

Reply:
Looked it up, and it prevents compiler RVO, as per the warning, will remove.